client/client-lib/src/query.rs

```diff
@@ -93,6 +93,61 @@ impl<R: Debug + Eq + Clone> QueryStrategy<Vec<R>> for UnionResponses<R> {
     }
 }
 
+/// Returns the deduplicated union of `required` number of responses
+///
+/// Unlike [`UnionResponses`], it works with single values, not `Vec`s.
+/// TODO: Should we make UnionResponses a wrapper around this one?
+pub struct UnionResponsesSingle<R> {
+    responses: HashSet<PeerId>,
+    existing_results: Vec<R>,
+    current: CurrentConsensus<Vec<R>>,
+    required: usize,
+}
+
+impl<R> UnionResponsesSingle<R> {
+    pub fn new(required: usize) -> Self {
+        Self {
+            responses: HashSet::new(),
+            existing_results: vec![],
+            current: CurrentConsensus::new(required),
+            required,
+        }
+    }
+}
+
+impl<R: Debug + Eq + Clone> QueryStrategy<R, Vec<R>> for UnionResponsesSingle<R> {
+    fn process(&mut self, response: FedResponse<R>) -> QueryStep<Vec<R>> {
+        match response {
+            FedResponse {
+                peer,
+                result: Ok(new_result),
+            } => {
+                if !self.existing_results.iter().any(|r| r == &new_result) {
```

Comment:
Forget why I didn't use a HashSet here...

Reply:
You did, and I converted it. :)

 There was some bound required on the whole thing that `HashSet` was breaking. Linear search on such a tiny structure is good enough.